src/System/Hatrace.hs

```diff
@@ -1442,36 +1449,267 @@ atomicWritesSink =
+formatHatraceEventConduit = CL.mapM $ \(pid, event) -> do
+  case event of
 
     SyscallStop enterOrExit -> case enterOrExit of
 
       SyscallEnter (syscall, syscallArgs) -> do
-        formatted <- getFormattedSyscallEnterDetails syscall syscallArgs pid
-        putStrLn $ show [pid] ++ " Entering syscall: " ++ show syscall
-          ++ (if formatted /= "" then ", details: " ++ formatted else "")
+        formatted <- liftIO $ formatSyscallEnter syscall syscallArgs pid
+        return $ HatraceEvent pid (EventSyscallEnter $ EventSyscallEnterDetails syscall formatted)
 
       SyscallExit (syscall, syscallArgs) -> do
-        formatted <- getFormattedSyscallExitDetails syscall syscallArgs pid
-        putStrLn $ show [pid] ++ " Exited syscall: " ++ show syscall
-          ++ (if formatted /= "" then ", details: " ++ formatted else "")
+        (formatted, outcome) <- liftIO $ formatSyscallExit syscall syscallArgs pid
+        return $ HatraceEvent pid (EventSyscallExit $ EventSyscallExitDetails syscall formatted outcome)
+
+    PTRACE_EVENT_Stop ptraceEvent ->
+      return $ HatraceEvent pid (EventPTraceEvent ptraceEvent)
+
+    GroupStop sig ->
+      return $ HatraceEvent pid (EventGroupStop sig)
+
+    SignalDeliveryStop sig ->
+      return $ HatraceEvent pid (EventSignalDelivery sig)
+
+    Death fullStatus ->
+      return $ HatraceEvent pid (EventProcessDeath fullStatus)
+
+printHatraceEvent :: StringFormattingOptions -> HatraceEvent -> IO ()
+printHatraceEvent formattingOptions (HatraceEvent pid details) = do
+  putStr $ show [pid] ++ " "
+  case details of
+    EventSyscallEnter enterDetails ->
+      let syscall = evEnterSyscall enterDetails
+          formattedSyscall = evEnterFormatted enterDetails
+      in putStrLn $ "Entering syscall: " ++ show syscall ++ ", details: " ++
+           syscallToString formattingOptions formattedSyscall
+
+    EventSyscallExit exitDetails ->
+      let syscall = evExitSyscall exitDetails
+          formattedSyscall = evExitFormatted exitDetails
+          outcome = case evExitOutcome exitDetails of
+            ProperReturn formattedReturn -> formattedReturn
+            ErrnoResult _errno strErr ->
+              FormattedReturn $ FixedStringArg $ "-1 (" ++ strErr ++ ")"
+      in putStrLn $ "Exited syscall: " ++ show syscall ++ ", details: " ++
+        syscallExitToString formattingOptions (formattedSyscall, outcome)
+
+    EventPTraceEvent ptraceEvent ->
+      putStrLn $ "Got event: " ++ show ptraceEvent
+
+    EventGroupStop sig ->
+      putStrLn $ "Got group stop: " ++ prettySignal sig
+
+    EventSignalDelivery sig ->
+      putStrLn $ "Got signal: " ++ prettySignal sig
+
+    EventProcessDeath exitCode ->
+      putStrLn $ "Process exited with status: " ++ show exitCode
+
+
+printHatraceEventJson :: HatraceEvent -> IO ()
+printHatraceEventJson hatraceEvent = do
+  BS.putStr $ BSL.toStrict $ encode hatraceEvent
+  BS.putStr "\n"
```

Comment:
Probably want to do that in 1 `putStr` syscall to reduce chances of interleaving.